### PR TITLE
Limit number of events in a replication request

### LINKED ISSUE #9956: Replication request exceeds max size when trying to join room
### Description

I'm getting the following error when trying to join #irc:matrix.org:

```
2021-05-07 20:16:08,373 - synapse.http.site - 107 - WARNING - sentinel - Aborting connection from IPv4Address(type='TCP', host='127.0.0.1', port=44100) because the request exceeds maximum size
2021-05-07 20:16:08,375 - synapse.http.site - 312 - INFO - sentinel - Connection from IPv4Address(type='TCP', host='127.0.0.1', port=44100) lost before request headers were read
2021-05-07 20:16:08,376 - synapse.http.client - 440 - INFO - POST-3 - Received response to POST http://127.0.0.1:9893/_synapse/replication/remote_join/%21BAXLHOFjvDKUeLafmO%3Amatrix.org/%40deepbluev7%3Aneko.dev/qjkxpFkuVk: 502
2021-05-07 20:16:08,376 - synapse.http.server - 88 - INFO - POST-3 - <SynapseRequest at 0x7f0f95586d00 method='POST' uri='/_matrix/client/r0/join/%23irc%3Amatrix.org' clientproto='HTTP/1.0' site=8086> SynapseError: 502 - Failed to talk to 
main process
2021-05-07 20:16:08,377 - synapse.access.http.8086 - 387 - INFO - POST-3 - 127.0.0.1 - 8086 - {@deepbluev7:neko.dev} Processed request: 27.573sec/-0.000sec (0.005sec, 0.001sec) (0.002sec/0.008sec/10) 64B 502 "POST /_matrix/client/r0/join/%
23irc%3Amatrix.org HTTP/1.0" "mtxclient v0.5.1" [0 dbevts]
```

I have a few workers with an event_creator and persister as well as a master and a few federation workers. This is from the event_creator log. Increasing the maximum size in https://github.com/matrix-org/synapse/blob/3ff225175462dde8376aa584e3a47c43b1f0e790/synapse/http/site.py#L106 made the join work.

### Steps to reproduce

- Setup synapse with event_creator
- Try to join #irc:matrix.org

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

### Version information

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**:

If not matrix.org: neko.dev

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**: 1.33.1

- **Install method**: custom ebuild
<!-- examples: package manager/git clone/pip  -->

- **Platform**: Gentoo
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes #9956

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -3056,8 +3056,9 @@ async def persist_events_and_notify(
         """
         instance = self.config.worker.events_shard_config.get_instance(room_id)
         if instance != self._instance_name:
-            # Limit the number of events sent over federation.
-            for batch in batch_iter(event_and_contexts, 1000):
+            # Limit the number of events sent over federation. We choose 200
```
Comment (richvdh) [inline synapse/handlers/federation.py:3059]:
```suggestion
            # Limit the number of events sent over replication. We choose 200
```

surely?

**Review by richvdh — APPROVED** [review]
lgtm otherwise

File: synapse/handlers/federation.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/federation.py:3059]: :facepalm:

File: changelog.d/10118.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix a bug causing replication requests to fail when receiving a lot of very large events via federation.
```
Comment (richvdh) [inline changelog.d/10118.bugfix:1]:
```suggestion
Fix a bug introduced in Synapse 1.33.0 which caused replication requests to fail when receiving a lot of very large events via federation.
```

**Pull request merged**